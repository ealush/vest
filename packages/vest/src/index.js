import any from "anyone/any";
import enforce from "n4s/dist/enforce.min";
import { VERSION } from "./constants";
import test from "./core/test";
import validate from "./core/validate";
import { draft, only, skip, warn } from "./hooks";
import { singleton, runWithContext } from "./lib";

export default singleton.register({
  Enforce: enforce.Enforce,
  VERSION,
  any,
  draft,
  enforce,
  only,
  runWithContext,
  skip,
  test,
  validate,
  warn,
});
