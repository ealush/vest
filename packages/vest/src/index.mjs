import enforce from '../../n4s/src/enforce';
import create from './core/suite/create';
import test from './core/test';
import { draft, only, skip, warn, group } from './hooks';

const VERSION = VEST_VERSION;

export default {
  VERSION,
  create,
  draft,
  enforce,
  group,
  only,
  skip,
  test,
  warn,
};

export { VERSION, create, draft, enforce, group, only, skip, test, warn };
