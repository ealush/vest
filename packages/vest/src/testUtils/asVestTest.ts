import { invariant } from 'vest-utils';

import { ErrorStrings } from 'ErrorStrings';
import { TIsolateTest } from 'IsolateTest';
import { isIsolateTest } from 'isIsolateTest';

export function asVestTest(value: any): TIsolateTest {
  invariant(isIsolateTest(value), ErrorStrings.EXPECTED_VEST_TEST);
  return value;
}
