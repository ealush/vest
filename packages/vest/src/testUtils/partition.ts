export default function partition<T>(
  array: T[],
  // eslint-disable-next-line no-unused-vars
  predicate: (value: T, index: number, array: T[]) => boolean
): [T[], T[]] {
  return array.reduce(
    (partitions: [T[], T[]], value, number) => {
      partitions[predicate(value, number, array) ? 0 : 1].push(value);
      return partitions;
    },
    [[], []]
  );
}
