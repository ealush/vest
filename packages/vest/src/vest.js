import create from 'createSuite';
import enforce from 'enforce';
import { only, skip, warn, group } from 'hooks';
import test from 'test';

const VERSION = __LIB_VERSION__;

export default {
  VERSION,
  create,
  enforce,
  group,
  only,
  skip,
  test,
  warn,
};
