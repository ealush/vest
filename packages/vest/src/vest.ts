import { enforce } from 'n4s';

import create from 'create';
import context from 'ctx';
import each from 'each';
import { only, skip } from 'exclusive';
import group from 'group';
import include from 'include';
import { eager } from 'mode';
import omitWhen from 'omitWhen';
import optional from 'optionalTests';
import skipWhen from 'skipWhen';
import { test } from 'test';
import warn from 'warn';

const VERSION = __LIB_VERSION__;
export {
  test,
  create,
  each,
  only,
  skip,
  warn,
  group,
  optional,
  skipWhen,
  omitWhen,
  enforce,
  VERSION,
  context,
  include,
  eager,
};
