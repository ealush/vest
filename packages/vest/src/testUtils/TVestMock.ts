import { TFieldName, TGroupName } from 'SuiteResultTypes';
import * as vest from 'vest';

export type TVestMock = typeof vest;

export type TTestSuiteCallback = (...args: any[]) => void;
export type TTestSuite = vest.Suite<TFieldName, TGroupName, TTestSuiteCallback>;
