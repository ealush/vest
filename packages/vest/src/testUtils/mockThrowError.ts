import { TVestMock } from './TVestMock';

export default function mockThrowError() {
  const deferThrow = jest.fn<string, any>();
  jest.resetModules();
  jest.mock('vest-utils', () => ({
    ...jest.requireActual('vest-utils'),
    deferThrow,
  }));
  const vest = require('vest') as TVestMock;

  return {
    deferThrow,
    vest,
  };
}
