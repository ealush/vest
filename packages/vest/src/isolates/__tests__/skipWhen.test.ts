import { TTestSuite } from 'testUtils/TVestMock';

import { dummyTest } from '../../testUtils/testDummy';

import * as vest from 'vest';

describe('skipWhen', () => {
  let fn = jest.fn();
  beforeEach(() => {
    fn = jest.fn();
    suite.reset();
  });
  it('Should run callback both when condition is true or false', () => {
    let counter = 0;
    const suite = vest.create(() => {
      vest.skipWhen(counter === 1, fn);

      counter++;
    });
    expect(fn).toHaveBeenCalledTimes(0);
    suite();
    expect(fn).toHaveBeenCalledTimes(1);
    suite();
    expect(fn).toHaveBeenCalledTimes(2);
  });

  it('Should respect both boolean and function conditions', () => {
    const suite = vest.create(() => {
      vest.skipWhen(false, fn);
      vest.skipWhen(true, fn);
      vest.skipWhen(() => false, fn);
      vest.skipWhen(() => true, fn);
    });

    suite();

    expect(fn).toHaveBeenCalledTimes(4);
  });

  it('Should pass result draft to the functional condition', () => {
    const f = jest.fn();
    const control = jest.fn();

    vest.create(() => {
      vest.skipWhen(draft => {
        expect(draft.hasErrors()).toBe(false);
        expect(draft).toMatchSnapshot();
        control();
        return false;
      }, f);
      dummyTest.failing('f1', 'msg');
      vest.skipWhen(draft => {
        expect(draft.hasErrors()).toBe(true);
        expect(draft.hasErrors('f1')).toBe(true);
        expect(draft.hasErrors('f2')).toBe(false);
        expect(draft.hasErrors('f3')).toBe(false);
        expect(draft).toMatchSnapshot();
        control();
        return false;
      }, f);
      dummyTest.failing('f2', 'msg');
      vest.skipWhen(draft => {
        expect(draft.hasErrors()).toBe(true);
        expect(draft.hasErrors('f1')).toBe(true);
        expect(draft.hasErrors('f2')).toBe(true);
        expect(draft.hasErrors('f3')).toBe(false);
        expect(draft).toMatchSnapshot();
        control();
        return false;
      }, f);
      dummyTest.failing('f3', 'msg');
    })();

    expect(control).toHaveBeenCalledTimes(3);
  });

  it('Should skip tests when the condition is truthy', () => {
    const res = suite(true);
    expect(res.tests.username.testCount).toBe(0);
  });

  it('Should run tests when the condition is falsy', () => {
    const res = suite(false);
    expect(res.tests.username.testCount).toBe(1);
  });

  it('Should correctly refill the state when field is skipped', () => {
    const res = suite(false);
    expect(res.tests.username.testCount).toBe(1);
    suite(true);

    expect(suite.get().tests.username.testCount).toBe(1);
  });

  describe('nested calls', () => {
    let suite: TTestSuite;

    describe('skipped in non-skipped', () => {
      beforeEach(() => {
        suite = vest.create(() => {
          vest.skipWhen(false, () => {
            vest.test('outer', () => false);

            vest.skipWhen(true, () => {
              vest.test('inner', () => false);
            });
          });
        });
        suite();
      });
      it('Should run `outer` and skip `inner`', () => {
        expect(suite.get().testCount).toBe(1);
        expect(suite.get().hasErrors('outer')).toBe(true);
        expect(suite.get().hasErrors('inner')).toBe(false);
      });
    });

    describe('skipped in skipped', () => {
      beforeEach(() => {
        suite = vest.create(() => {
          vest.skipWhen(true, () => {
            vest.test('outer', () => false);

            vest.skipWhen(true, () => {
              vest.test('inner', () => false);
            });
          });
        });
        suite();
      });
      it('Should skip both `outer` and `inner`', () => {
        expect(suite.get().testCount).toBe(0);
        expect(suite.get().hasErrors('outer')).toBe(false);
        expect(suite.get().hasErrors('inner')).toBe(false);
      });
    });
    describe('non-skipped in skipped', () => {
      beforeEach(() => {
        suite = vest.create(() => {
          vest.skipWhen(true, () => {
            vest.test('outer', () => false);

            vest.skipWhen(false, () => {
              vest.test('inner', () => false);
            });
          });
        });
        suite();
      });
      it('Should skip both', () => {
        expect(suite.get().testCount).toBe(0);
        expect(suite.get().hasErrors('outer')).toBe(false);
        expect(suite.get().hasErrors('inner')).toBe(false);
      });
    });
  });
});

const suite = vest.create((skipTest: boolean) => {
  vest.skipWhen(skipTest, () => {
    vest.test('username', () => false);
  });
  vest.test('control', () => false);
});
