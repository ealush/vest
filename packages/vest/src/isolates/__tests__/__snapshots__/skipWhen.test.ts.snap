// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`skipWhen Should pass result draft to the functional condition 1`] = `
SuiteSummary {
  "errorCount": 0,
  "errors": [],
  "getError": [Function],
  "getErrors": [Function],
  "getErrorsByGroup": [Function],
  "getWarning": [Function],
  "getWarnings": [Function],
  "getWarningsByGroup": [Function],
  "groups": {},
  "hasErrors": [Function],
  "hasErrorsByGroup": [Function],
  "hasWarnings": [Function],
  "hasWarningsByGroup": [Function],
  "isValid": [Function],
  "isValidByGroup": [Function],
  "pendingCount": 0,
  "suiteName": undefined,
  "testCount": 0,
  "tests": {},
  "valid": false,
  "warnCount": 0,
  "warnings": [],
}
`;

exports[`skipWhen Should pass result draft to the functional condition 2`] = `
SuiteSummary {
  "errorCount": 1,
  "errors": [
    SummaryFailure {
      "fieldName": "f1",
      "groupName": undefined,
      "message": "msg",
    },
  ],
  "getError": [Function],
  "getErrors": [Function],
  "getErrorsByGroup": [Function],
  "getWarning": [Function],
  "getWarnings": [Function],
  "getWarningsByGroup": [Function],
  "groups": {},
  "hasErrors": [Function],
  "hasErrorsByGroup": [Function],
  "hasWarnings": [Function],
  "hasWarningsByGroup": [Function],
  "isValid": [Function],
  "isValidByGroup": [Function],
  "pendingCount": 0,
  "suiteName": undefined,
  "testCount": 1,
  "tests": {
    "f1": SummaryBase {
      "errorCount": 1,
      "errors": [
        "msg",
      ],
      "pendingCount": 0,
      "testCount": 1,
      "valid": false,
      "warnCount": 0,
      "warnings": [],
    },
  },
  "valid": false,
  "warnCount": 0,
  "warnings": [],
}
`;

exports[`skipWhen Should pass result draft to the functional condition 3`] = `
SuiteSummary {
  "errorCount": 2,
  "errors": [
    SummaryFailure {
      "fieldName": "f1",
      "groupName": undefined,
      "message": "msg",
    },
    SummaryFailure {
      "fieldName": "f2",
      "groupName": undefined,
      "message": "msg",
    },
  ],
  "getError": [Function],
  "getErrors": [Function],
  "getErrorsByGroup": [Function],
  "getWarning": [Function],
  "getWarnings": [Function],
  "getWarningsByGroup": [Function],
  "groups": {},
  "hasErrors": [Function],
  "hasErrorsByGroup": [Function],
  "hasWarnings": [Function],
  "hasWarningsByGroup": [Function],
  "isValid": [Function],
  "isValidByGroup": [Function],
  "pendingCount": 0,
  "suiteName": undefined,
  "testCount": 2,
  "tests": {
    "f1": SummaryBase {
      "errorCount": 1,
      "errors": [
        "msg",
      ],
      "pendingCount": 0,
      "testCount": 1,
      "valid": false,
      "warnCount": 0,
      "warnings": [],
    },
    "f2": SummaryBase {
      "errorCount": 1,
      "errors": [
        "msg",
      ],
      "pendingCount": 0,
      "testCount": 1,
      "valid": false,
      "warnCount": 0,
      "warnings": [],
    },
  },
  "valid": false,
  "warnCount": 0,
  "warnings": [],
}
`;
